modules/ml/src/main/java/org/apache/ignite/ml/knn/classification/KNNClassificationModel.java

```diff
@@ -17,145 +17,164 @@
+     * Constructs a new instance of KNN classification model.
      *
-     * @param dataset Specially prepared object to run algorithm over it.
+     * @param dataset Dataset with {@link SpatialIndex} as a partition data.
+     * @param distanceMeasure Distance measure.
+     * @param k Number of neighbours.
+     * @param weighted Weighted or not.
      */
-    public KNNClassificationModel(Dataset<EmptyContext, LabeledVectorSet<Double, LabeledVector>> dataset) {
-        this.datasets = new ArrayList<>();
-        if (dataset != null)
-            datasets.add(dataset);
+    KNNClassificationModel(Dataset<EmptyContext, SpatialIndex<Double>> dataset, DistanceMeasure distanceMeasure, int k,
+        boolean weighted) {
+        super(dataset, distanceMeasure, k, weighted);
     }
 
     /** {@inheritDoc} */
-    @Override public Double predict(Vector v) {
-        if (!datasets.isEmpty()) {
-            List<LabeledVector> neighbors = findKNearestNeighbors(v);
+    @Override public Double predict(Vector input) {
+        List<LabeledVector<Double>> neighbors = findKClosest(k, input);
 
-            return classify(neighbors, v, stgy);
-        }
-        else
-            throw new IllegalStateException("The train kNN dataset is null");
+        return election(neighbors, input);
     }
 
-    /** */
-    @Override public <P> void saveModel(Exporter<KNNModelFormat, P> exporter, P path) {
-        KNNModelFormat mdlData = new KNNModelFormat(k, distanceMeasure, stgy);
-        exporter.save(mdlData, path);
+    /**
+     * Elects a label with max votes for it.
+     *
+     * @param neighbours List of neighbours with different labels.
+     * @param pnt Point to calculate distance to.
+     * @return Label with max votes for it.
+     */
+    private Double election(List<LabeledVector<Double>> neighbours, Vector pnt) {
+        Collection<GroupedNeighbours> groups = groupByLabel(neighbours);
+
+        return election(groups, pnt);
     }
 
     /**
-     * The main idea is calculation all distance pairs between given vector and all vectors in training set, sorting
-     * them and finding k vectors with min distance with the given vector.
+     * Elects a label with max votes for it.
      *
-     * @param v The given vector.
-     * @return K-nearest neighbors.
+     * @param groups Groups of neighbours (each group contains neighbours with the same label).
+     * @param pnt Point to calculate distance to.
+     * @return Label with max votes for it.
      */
-    protected List<LabeledVector> findKNearestNeighbors(Vector v) {
-        List<LabeledVector> neighborsFromPartitions = datasets.stream()
-            .flatMap(dataset -> findKNearestNeighborsInDataset(v, dataset).stream())
-            .collect(Collectors.toList());
-
-        LabeledVectorSet<Double, LabeledVector> neighborsToFilter = buildLabeledDatasetOnListOfVectors(neighborsFromPartitions);
+    private Double election(Collection<GroupedNeighbours> groups, Vector pnt) {
```

Comment:
Could you please exaplain the main idea of this refactoring?

Reply:
A bit more clarity I guess. I actually have written this class from scratch, so refactoring itself wasn't my first purpose.